cnmodel/util/sound.py

```diff
@@ -495,7 +458,77 @@ def generate(self):
         basenoise = pipnoise(self.time, o['ramp_duration'], o['rate'],
                         o['dbspl'], o['pip_duration'], o['pip_start'], o['seed'])
         return sinusoidal_modulation(self.time, basenoise, o['pip_start'], o['fmod'], o['dmod'], 0.)
+
+
+class ReadWavefile(Sound):
+    """ Read a .wav file from disk, possibly converting the sample rate and the scale
+    for use in driving the auditory nerve fiber model.
+    
+    Parameters
+    ----------
+    wavefile : str
+        name of the .wav file to read
+    rate : float
+        Sample rate in Hz (waveform will be resampled to this rate)
+    dbspl : float or None
+        If specified, the wave file is scaled such that its overall dBSPL
+        (measured from RMS of the entire waveform) is equal to this value.
+        Either ``dbspl`` or ``scale`` must be specified.
+    scale : float or None
+        If specified, the wave data is multiplied by this value to yield values in dBSPL. 
+        Either ``dbspl`` or ``scale`` must be specified.
+    delay: float (default: 0.)
+        delay time to start sound, in s. Allows anmodel and cells to run to steady-state
+    maxdur : float or None (default: None)
+        If not None, then sets maximum duration of waveform to return (in seconds).
+        Otherwise, if None, sets the duration of the waveform from the wavefile.
+    channel: int (default: 0)
+        If wavefile has 2 channels, select 0 or 1 for the channel to read
+    
+    Returns
+    -------
+    array :
+        waveform
     
+    """
+    def __init__(self, wavefile, rate, dbspl=None, scale=None, maxdur=None, channel=0, delay=0.):
+        if dbspl is not None and scale is not None:
+            raise ValueError('Only one of "dbspl" or "scale" can be set')
+        duration = 0.  # forced because of the way num_samples has to be calculated first
+        Sound.__init__(self, duration, rate, dbspl=dbspl, scale=scale, wavefile=wavefile,
+                maxdur=maxdur, channel=channel, delay=delay)
+        
+    def generate(self):
+        """
+        Read the wave file from disk, clip duration, resample if necessary, and scale
+        
+        Returns
+        -------
+        array :   generated waveform
+        """
+        [fs_wav, stimulus] = scipy.io.wavfile.read(self.opts['wavefile']) # raw is a numpy array of integer, representing the samples
+        if len(stimulus.shape) > 1 and stimulus.shape[1] > 0:
+            stimulus = stimulus[:,self.opts['channel']]  # just use selected channel
+
+        fs_wav = float(fs_wav)
+        wavedur = len(stimulus)/fs_wav 
+        maxdur = self.opts['maxdur']
+        if maxdur is not None and wavedur > maxdur:  # clip waveform to specified
+            st = np.zeros(int(maxdur*fs_wav))
+            st[int(self.opts['delay']*fs_wav):int(maxdur*fs_wav)] = stimulus[0:int((maxdur-self.opts['delay'])*fs_wav)]  # prepend a silent startup delay
```

Comment:
delay needs to be handled outside this conditional.

Reply:
Fixed - Now, if maxdur is None, the resulting waveform is the original waveform with a silent delay (may have duration of 0) inserted at the beginning. Otherwise, the waveform has a total duration of maxdur (including the delay).